README.md
# Phusion Passenger APT repository automation tools

This repository contains tools for automatically creating a multi-distribution APT repository for Phusion Passenger. The goal is to automatically build Debian packages for multiple distributions, immediately after a source release. These tools are meant to be run on Ubuntu 12.04.

First, install prerequisites and setup a user:

    sudo apt-get install ubuntu-dev-tools reprepro debhelper source-highlight ruby1.9.3
    sudo gem install bluecloth mizuho drake --no-rdoc --no-ri
    sudo adduser psg_apt_automation

Add this to /etc/sudoers:

    Cmnd_Alias PBUILDER_CREATE = /usr/sbin/pbuilder --create *
    Cmnd_Alias PBUILDER_UPDATE = /usr/sbin/pbuilder --update *
    Cmnd_Alias PBUILDER_BUILD = /usr/sbin/pbuilder --build *
    Cmnd_Alias PBUILDER=PBUILDER_CREATE,PBUILDER_UPDATE,PBUILDER_BUILD
    Defaults!PBUILDER env_keep="ARCHITECTURE DISTRIBUTION ARCH DIST DEB_BUILD_OPTIONS HOME"
    psg_apt_automation ALL=(root)NOPASSWD:PBUILDER

Now install the pbuilder distributions. You need to patch pbuilder-dist and add Debian Wheezy support first. Edit `/usr/lib/pbuilder/pbuilder-apt-config` and replace:

    etch|lenny|squeeze|sid|oldstable|stable|testing|unstable|experimental)

with:

    etch|lenny|squeeze|sid|oldstable|stable|testing|unstable|experimental|wheezy)

Then run:

    sudo -u psg_apt_automation -H ./setup-pbuilder-dist

Configure your GPG signing settings:

    sudo nano configuration   # setup your key ID
    sudo nano passphrase      # put your password here
    sudo chown psg_apt_automation:psg_apt_automation passphrase
    sudo chmod 600 passphrase
    sudo -u psg_apt_automation -H gpg --keyserver keyserver.ubuntu.com --recv-keys C324F5BB38EEB5A0
    sudo -u psg_apt_automation -H gpg --armor --export C324F5BB38EEB5A0 | sudo apt-key add -

Import miscellaneous Phusion packages:

    ./import_misc_packages

Then, every time a new Phusion Passenger version is released, run the following command to update the APT repository in `apt/`, as `psg_apt_automation`:

    ./new_release <GIT_URL> <REPO_DIR> <APT_DIR> [REF]

where:

 * `GIT_URL` is the Phusion Passenger git repository URL.
 * `REPO_DIR` is the directory that the git repository should be cloned to.
 * `APT_DIR` is the APT repository directory.
 * `REF` is the commit in git for which you want to build packages. If `REF` is not specified, then it is assumed to be `origin/master`.

For example:

    ./new_release https://github.com/phusion/passenger.git passenger.repo passenger.apt

The `new_release` script is near-atomic: it is very unlikely that users will see an intermediate state in which only some packages have been built.

## Related project

 * https://github.com/phusion/passenger_autobuilder


nginx-debian/changelog
nginx (1.4.1-3ubuntu1) saucy; urgency=low

  * Resynchronise with Debian.  Remaining changes:
    - debian/patches/ubuntu-branding.patch:
      + Add Ubuntu branding to server_tokens.

 -- Colin Watson <cjwatson@ubuntu.com>  Thu, 20 Jun 2013 15:08:44 +0100

nginx (1.4.1-3) unstable; urgency=low

  [ Kartik Mistry ]
  * debian/control:
    + Changed libgd2-dev build-dep to libgd2-dev|libgd2-noxpm-dev allow
      backporting (Closes: #711505)

  [ Cyril Lavier ]
  * debian/nginx-naxsi-ui.preinst
    + Added argument "install" to the preinst script. (Closes: #711590)

 -- Kartik Mistry <kartik@debian.org>  Sun, 09 Jun 2013 13:07:52 +0530

nginx (1.4.1-2ubuntu1) saucy; urgency=low

  * Resynchronise with Debian.  Remaining changes:
    - debian/patches/ubuntu-branding.patch:
      + Add Ubuntu branding to server_tokens.

 -- Colin Watson <cjwatson@ubuntu.com>  Fri, 07 Jun 2013 15:41:48 +0100

nginx (1.4.1-2) unstable; urgency=medium

  [ Kartik Mistry]
  * Urgency set to medium due to Security and RC bugs.
  * debian/conf/sites-available/default:
    + Fixed typo in listen ipv6only=on (Closes: #707684)

  [ Cyril Lavier ]
  * debian/control:
    + Replaced the build-dep libgd2-noxpm-dev to libgd2-dev as part of
      the libgd2-{xpm,noxpm}-dev -> libgd2-dev transition.
  * debian/nginx-naxsi-ui.preinst, debian/nginx-naxsi-ui.postinst,
    debian/conf/naxsi-ui.conf.1.4.1:
    + Added preinst script and renamed the new config file to avoid
      prompting the user. (Closes: #707291)
  * debian/rules:
    + Made "debian/rules binary" enough to build binary package.
      (Closes: #708522)

 -- Kartik Mistry <kartik@debian.org>  Thu, 06 Jun 2013 12:55:23 +0530

nginx (1.4.1-1ubuntu2) saucy; urgency=low

  * Rebuild for libgd3.

 -- Colin Watson <cjwatson@ubuntu.com>  Mon, 20 May 2013 00:45:10 +0100

nginx (1.4.1-1ubuntu1) saucy; urgency=low

  * Merge with Debian unstable (LP: #1177919). Remaining changes:
    - debian/conf/sites-available/default:
      + Modify default site configuration file to correct a typo
        that prevented out-of-the-box usability (LP: #1162177).
    - debian/patches/ubuntu-branding.patch:
      + Add ubuntu branding to server_tokens.
  * Refresh all patches.

 -- Dmitry Shachnev <mitya57@ubuntu.com>  Sat, 11 May 2013 14:47:53 +0400

nginx (1.4.1-1) unstable; urgency=low

  * New upstream release:
    + Fixes arbitrary code execution (CVE-2013-2028).
  * Uploaded to unstable.
  * debian/control:
    + Updated Standards-Version to 3.9.4

 -- Kartik Mistry <kartik@debian.org>  Tue, 07 May 2013 19:53:46 +0530

nginx (1.4.0-2) experimental; urgency=low

  [ Ondřej Surý ]
  * debian/modules/:
    + Updated nginx-echo, nginx-cache-purge and naxsi modules.
    + Removed useless .gitignore, .gitmodules files.

  [ Kartik Mistry ]
  * debian/modules/nginx-upload:
    + This module no longer works with 1.3.x and above. Removed as of now.
  * debian/modules/ngx-fancyindex:
    + Added Fancy Indexes module (Closes: #704210)
  * debian/copyright:
    + Fixed path for modules in Files: field.
    + Updated copyright for debian/*
  * debian/rules:
    + Enabled spdy module (Closes: #706195).
  * debian/control:
    + Suggests: fcgiwrap (Closes: #701508).

  [ Cyril Lavier ]
  * debian/conf/naxsi-ui.conf:
    + Added configuration file for nginx-naxsi-ui using SQLite
      (Closes: #699678).
  * debian/nginx-naxsi-ui.config, debian/nginx-naxsi-ui.postinst,
    debian/nginx-naxsi-ui.postrm, debian/nginx-naxsi-ui.prerm,
    debian/nginx-naxsi-ui.templates:
    + Removed these files as they are not necessary anymore with the
      database engine switching from MySQL to SQLite.
  * debian/control:
    + Removed dependencies against MySQL.
  * debian/rules, debian/modules/ngx_http_substitutions_filter_module,
    debian/README.Modules-versions:
    + Added http_substitutions_filter module. (Closes: #706456)

 -- Kartik Mistry <kartik@debian.org>  Wed, 01 May 2013 10:48:43 +0530

nginx (1.4.0-1) experimental; urgency=low

  [ Kartik Mistry ]
  * New upstream release (Closes: #706127).
  * debian/rules, debian/modules/:
    + Removed chunkin-nginx-module as it no longer supported for nginx 1.3.9+
    + Updated nginx-lua module to 0.8.0
    + Patched nginx-upload module as described in modules/README file.
  * Refreshed debian/patches/perl-use-dpkg-buildflags.patch
  * debian/logrotate:
    + Set default log to keep for 52 weeks instead of 52 days. Thanks to
      RjY <rjy@users.sourceforge.net> for patch (Closes: #696440)
  * debian/rules:
    + Added cache purge module to nginx-extras.
  * debian/control:
    + Suggests: nginx-docs. Thanks to colliar <colliar4ever@aol.com> for
      reporting bug (Closes: #702923)
  * debian/copyright:
    + Updated copyright year.
    + Fixed broken license text.

  [ Michael Lustfield ]
  * debian/conf/sites-available/default:
    + Added ipv6_only=on to default server block. (Closes: #700857)
    + Added default_server to ipv4 default server block. Now matches ipv6.
  * debian/rules:
    + Removed obsolete --with-md5 and --with-sha1

 -- Kartik Mistry <kartik@debian.org>  Thu, 25 Apr 2013 12:51:45 +0530

nginx (1.2.6-1ubuntu4) saucy; urgency=low

  * debian/patches/ubuntu-branding.patch: Move Ubuntu branding from
    the numerical version string to the long version (LP: #1174158)

 -- Adam Conrad <adconrad@ubuntu.com>  Thu, 09 May 2013 11:50:52 -0600

nginx (1.2.6-1ubuntu3) raring; urgency=low

  * debian/patches/ubuntu-branding.patch: Add ubuntu branding
    to server_tokens. 

 -- Chuck Short <zulcss@ubuntu.com>  Tue, 09 Apr 2013 10:23:26 -0500

nginx (1.2.6-1ubuntu2) raring; urgency=low

  * debian/conf/sites-available/default:
    * Modify default site configuration file to correct a typo
      that prevented out-of-the-box usability (LP: #1162177)

 -- Thomas Ward <teward@ubuntu.com>  Fri, 05 Apr 2013 20:22:38 +0100

nginx (1.2.6-1ubuntu1) raring; urgency=low

  * debian/conf/sites-available/default:

preprocessor.rb
# encoding: utf-8
#  Phusion Passenger - https://www.phusionpassenger.com/
#  Copyright (c) 2013 Phusion
#
#  "Phusion Passenger" is a trademark of Hongli Lai & Ninh Bui.
#
#  Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to deal
#  in the Software without restriction, including without limitation the rights
#  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
#  copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#
#  The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
#
#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
#  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
#  THE SOFTWARE.

# Implements a simple preprocessor language which combines elements in the C
# preprocessor with ERB:
# 
#     Today
#     #if @today == :fine
#         is a fine day.
#     #elif @today == :good
#         is a good day.
#     #else
#         is a sad day.
#     #endif
#     Let's go walking.
#     Today is <%= Time.now %>.
# 
# When run with...
# 
#     Preprocessor.new.start('input.txt', 'output.txt', :today => :fine)
# 
# ...will produce:
# 
#     Today
#     is a fine day.
#     Let's go walking.
#     Today is 2013-08-11 22:37:06 +0200.
# 
# Highlights:
# 
#  * #if blocks can be nested.
#  * Expressions are Ruby expressions, evaluated within the binding of a
#    Preprocessor::Evaluator object.
#  * Text inside #if/#elif/#else are automatically unindented.
#  * ERB compatible.
class Preprocessor
	def initialize
		require 'erb' if !defined?(ERB)
		@indentation_size = 4
		@debug = boolean_option('DEBUG')
	end

	def start(filename, output_filename, variables = {})
		if output_filename
			temp_output_filename = "#{output_filename}._new"
			output = File.open(temp_output_filename, 'w')
		else
			output = STDOUT
		end
		the_binding  = create_binding(variables)
		context      = []
		@filename    = filename
		@lineno      = 1
		@indentation = 0

		each_line(filename) do |line|
			debug("context=#{context.inspect}, line=#{line.inspect}")

			name, args_string, cmd_indentation = recognize_command(line)
			case name
			when "if"
				case context.last
				when nil, :if_true, :else_true
					check_indentation(cmd_indentation)
					result = the_binding.eval(args_string, filename, @lineno)
					context.push(result ? :if_true : :if_false)
					inc_indentation
				when :if_false, :else_false, :if_ignore
					check_indentation(cmd_indentation)
					inc_indentation
					context.push(:if_ignore)
				else
					terminate "#if is not allowed in this context"
				end
			when "elif"
				case context.last
				when :if_true
					dec_indentation
					check_indentation(cmd_indentation)
					inc_indentation
					context[-1] = :if_false
				when :if_false
					dec_indentation
					check_indentation(cmd_indentation)
					inc_indentation
					result = the_binding.eval(args_string, filename, @lineno)
					context[-1] = result ? :if_true : :if_false
				when :else_true, :else_false
					terminate "#elif is not allowed after #else"
				when :if_ignore
					dec_indentation
					check_indentation(cmd_indentation)
					inc_indentation
				else
					terminate "#elif is not allowed outside #if block"
				end
			when "else"
				case context.last
				when :if_true
					dec_indentation
					check_indentation(cmd_indentation)
					inc_indentation
					context[-1] = :else_false
				when :if_false
					dec_indentation
					check_indentation(cmd_indentation)
					inc_indentation
					context[-1] = :else_true
				when :else_true, :else_false
					terminate "it is not allowed to have multiple #else clauses in one #if block"
				when :if_ignore
					dec_indentation
					check_indentation(cmd_indentation)
					inc_indentation
				else
					terminate "#else is not allowed outside #if block"
				end
			when "endif"
				case context.last
				when :if_true, :if_false, :else_true, :else_false, :if_ignore
					dec_indentation
					check_indentation(cmd_indentation)
					context.pop
				else
					terminate "#endif is not allowed outside #if block"
				end
			when "DEBHELPER"
				output.puts(line)
			when "", nil
				# Either a comment or not a preprocessor command.
				case context.last
				when nil, :if_true, :else_true
					output.puts(unindent(line))
				else
					# Check indentation but do not output.
					unindent(line)
				end
			else
				terminate "Unrecognized preprocessor command ##{name.inspect}"
			end

			@lineno += 1
		end
	ensure
		if output_filename && output
			output.close
			stat = File.stat(filename)
			File.chmod(stat.mode, temp_output_filename)
			File.chown(stat.uid, stat.gid, temp_output_filename) rescue nil
			File.rename(temp_output_filename, output_filename)
		end
	end

private
	UBUNTU_DISTRIBUTIONS = {
		"lucid"    => "10.04",
		"maverick" => "10.10",
		"natty"    => "11.04",
		"oneiric"  => "11.10",
		"precise"  => "12.04",
		"quantal"  => "12.10",
		"raring"   => "13.04",
		"saucy"    => "13.10"
	}

	# Provides the DSL that's accessible within.
	class Evaluator
		def _infer_distro_table(name)
			if UBUNTU_DISTRIBUTIONS.has_key?(name)
				return UBUNTU_DISTRIBUTIONS
			end
		end

		def is_distribution?(expr)
			if @distribution.nil?
				raise "The :distribution variable must be set"
			else
				if expr =~ /^(>=|>|<=|<|==|\!=)[\s]*(.+)/
					comparator = $1
					name = $2
				else
					raise "Invalid expression #{expr.inspect}"
				end

				table1 = _infer_distro_table(@distribution)
				table2 = _infer_distro_table(name)
				raise "Distribution name #{@distribution.inspect} not recognized" if !table1
				raise "Distribution name #{name.inspect} not recognized" if !table2
				v1 = table1[@distribution]
				v2 = table2[name]
				
				case comparator
				when ">"
					return v1 > v2
				when ">="
					return v1 >= v2
				when "<"
					return v1 < v2
				when "<="
					return v1 <= v2
				when "=="
					return v1 == v2
				when "!="
					return v1 != v2
				else
					raise "BUG"
				end
			end
		end
	end

	def each_line(filename)
		data = File.open(filename, 'r') do |f|
			erb = ERB.new(f.read, nil, "-")
			erb.filename = filename
			erb.result(binding)
		end
		data.each_line do |line|
			yield line.chomp
		end
	end
	
	def recognize_command(line)
		if line =~ /^([\s\t]*)#(.+)/
			indentation_str = $1
			command = $2

			# Declare tabs as equivalent to 4 spaces. This is necessary for
			# Makefiles in which the use of tabs is required.
			indentation_str.gsub!("\t", "    ")

			name = command.scan(/^\w+/).first
			# Ignore shebangs and comments.
			return if name.nil?

			args_string = command.sub(/^#{Regexp.escape(name)}[\s\t]*/, '')
			return [name, args_string, indentation_str.to_s.size]
		else
			return nil
		end
	end

	def create_binding(variables)
		object = Evaluator.new
		variables.each_pair do |key, val|
			object.send(:instance_variable_set, "@#{key}", val)
		end
		return object.instance_eval do
			binding
		end
	end

	def inc_indentation
		@indentation += @indentation_size
	end

	def dec_indentation
		@indentation -= @indentation_size
	end

	def check_indentation(expected)
		if expected != @indentation
			terminate "wrong indentation: found #{expected} characters, should be #{@indentation}"
		end
	end

	def unindent(line)
		line =~ /^([\s\t]*)/
		# Declare tabs as equivalent to 4 spaces. This is necessary for
		# Makefiles in which the use of tabs is required.
		found = $1.to_s.gsub("\t", "    ").size
		
		if found >= @indentation
			# Tab-friendly way to remove indentation.
			remaining = @indentation
			line = line.dup
			while remaining > 0
				if line[0..0] == " "
					remaining -= 1
				else
					# This is a tab.
					remaining -= 4
				end
				line.slice!(0, 1)
			end
			return line
		else
			terminate "wrong indentation: found #{found} characters, should be at least #{@indentation}"
		end
	end

	def debug(message)
		puts "DEBUG:#{@lineno}: #{message}" if @debug
	end

	def terminate(message)
		abort "*** ERROR: #{@filename} line #{@lineno}: #{message}"
	end
end
